### PR TITLE
Initial codebase for Storage component

### PR DESCRIPTION
**Pull request by thngkaiyuan**
## Example.java - An example of using Storage
Note: Remember to change the directory path to a valid path on your computer.
```
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.TreeSet;

import jfdi.storage.FileStorage;
import jfdi.storage.Storage;
import jfdi.storage.exceptions.ExistingFilesFoundException;
import jfdi.storage.exceptions.FilePathPair;
import jfdi.storage.exceptions.InvalidPathException;
import jfdi.storage.records.Alias;
import jfdi.storage.records.Task;

public class Example {

	public static void main(String[] args) {
		Storage s = FileStorage.getInstance();
		try {
			s.load("C:\\Users\\Kai Yuan\\Desktop\\JFDI");
		} catch (InvalidPathException e) {
			e.printStackTrace();
		} catch (ExistingFilesFoundException e) {
			for (FilePathPair pair : e.getReplacedFilePairs()) {
				System.out.println(pair.getOldFilePath() + " to "
						+ pair.getNewFilePath());
			}
			e.printStackTrace();
		}

		System.out.println(Task.getFilePath().toString());

		Task task = new Task();
		task.setDescription("test description");
		task.setStartDateTime(LocalDateTime.now());
		task.setEndDateTime(LocalDateTime.MAX);

		HashSet<String> tags = new HashSet<String>();
		tags.add("tag1");
		tags.add("tag2");
		task.setTags(tags);

		TreeSet<Duration> reminders = new TreeSet<Duration>();
		Duration duration1 = Duration.ZERO.plusDays(1);
		Duration duration2 = Duration.ZERO.plusDays(2);
		reminders.add(duration1);
		reminders.add(duration2);
		task.setReminders(reminders);

		Task task2 = new Task();
		task2.setDescription("test2 description");
		task2.setStartDateTime(LocalDateTime.now());
		task2.setEndDateTime(LocalDateTime.MAX);

		HashSet<String> tags2 = new HashSet<String>();
		tags2.add("tag3");
		tags2.add("tag4");
		task2.setTags(tags2);

		TreeSet<Duration> reminders2 = new TreeSet<Duration>();
		Duration duration12 = Duration.ZERO.plusDays(3);
		Duration duration22 = Duration.ZERO.plusDays(4);
		reminders2.add(duration12);
		reminders2.add(duration22);
		task2.setReminders(reminders2);

		System.out.println(task.createAndPersist());
		System.out.println(task.getId());
		System.out.println(task2.createAndPersist());
		System.out.println(task2.getId());
		System.out.println(Task.destroy(3));
		System.out.println(Task.markAsComplete(1));
		Task.persist();
		System.out.println(Task.destroy(2));
		System.out.println(Task.load());

		Alias alias1 = new Alias("hihi", "hoho");
		System.out.println(alias1.createAndPersist());
		Alias alias2 = new Alias("heyhey", "hoho");
		System.out.println(alias2.createAndPersist());
	}

}
```

## Todo
- [ ] Add interfaces for `Task` and `Alias`
- [ ] Prefix interfaces with `I`
- [ ] Refactor code
- [ ] Add tests

### CODE REVIEW

File: src/main/java/jfdi/storage/serializer/DurationDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+
+public class DurationDeserializer implements JsonDeserializer<Duration> {
+
+    @Override
+    public Duration deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+            throws JsonParseException {
+        String readableDurationString = json.getAsJsonPrimitive().getAsString();
+        String durationString = "PT" + readableDurationString;
```
Comment (thngkaiyuan) [inline src/main/java/jfdi/storage/serializer/DurationDeserializer.java]: Magic string?

**Comment by xinan** [comment]
Would it be better to have interfaces defined for the APIs and then your classes implement the interfaces?

**Comment by thngkaiyuan** [comment]
Hmmm u mean for Task and Alias? Yeah I think I may eventually have to do that so that you guys can replace them with stubs..

**Comment by xinan** [comment]
And for `Storage` as well :stuck_out_tongue:

**Comment by thngkaiyuan** [comment]
Storage is already an interface :)

**Comment by xinan** [comment]
:flushed: Oops. Maybe you want to name it `IStorage`. When I look at the name I thought it was a class.

**Comment by thngkaiyuan** [comment]
Ahh I wanted to do that.. But then I didn't see that kind of naming convention elsewhere so I thought maybe I should just name it normally.. So perhaps I will have `IStorage`, `ITask`, and `IAlias`? Or should I be more verbose by naming them `StorageInterface`, `TaskInterface` and `AliasInterface`?

**Comment by xinan** [comment]
I think `IStorage` would be better. `StorageInterface` is too long. Anyway I saw HubTurbo using `IModel`. But it have some interfaces without `I` also :slightly_frowning_face:

**Pull request closed without merging**